[derekjxtan] iP

derekjxtan opened the pull request:
# DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2324s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
##
If you Java programmer, you can use it to practice Java too. Here's the main method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

derekjxtan committed:
completed level-0

derekjxtan committed:
completed level-1

derekjxtan committed:
completed level-2

derekjxtan committed:
completed level 3

derekjxtan committed:
level 4 completed

derekjxtan committed:
completed TextUiTesting

derekjxtan committed:
completed level 5

derekjxtan committed:
completed level 6

derekjxtan committed:
refactored Duke class

derekjxtan committed:
Added invalidCommand exceptions for mark unmark and delete

derekjxtan committed:
Add Enums

derekjxtan committed:
Add load state from file functionality

derekjxtan committed:
Update Task to a abstract class

derekjxtan committed:
Task class: Add abstract function toSaveStateString

derekjxtan committed:
Task class: Add getters for done and task name

derekjxtan committed:
ToDo class: Add implementation of toSaveStateString function

derekjxtan committed:
Deadline class: Add implementation of toSaveStateString function

derekjxtan committed:
Event class: Add implementation of toSaveStateString function

derekjxtan committed:
Duke class: Implement saveState function

derekjxtan committed:
bug fix: Edit Deadline and Event class to save with the correct task type

derekjxtan committed:
Duke class: Add log to indicate state has been successfully loaded

derekjxtan committed:
Merge branch 'branch-level-7'

derekjxtan committed:
Deadline class: Change by attribute type from String to LocalDateTime

derekjxtan committed:
Event class: Change from and to attributes type from String to LocalDateTime

derekjxtan committed:
Task class: Add abstract function isOnDate

derekjxtan committed:
Deadline class: Add isOnDate implementation

derekjxtan committed:
Event class: Add isOnDate implementation

derekjxtan committed:
Duke class: Add on command functionality to list deadlines/events on a date

derekjxtan committed:
Merge branch 'branch-level-8'

derekjxtan committed:
Parser class: Add class skeleton

derekjxtan committed:
Parser class: Add implementation for parseUserInput

derekjxtan committed:
Parser class: Update parseUserInput to throw InvalidCommandException

derekjxtan committed:
Parser class: Add parseFileInput function

derekjxtan committed:
Duke class: Update main function to use Parser class

derekjxtan committed:
Add TaskList class

derekjxtan committed:
TaskList class: Add add task functionality

derekjxtan committed:
TaskList class: Add remove task functionality

derekjxtan committed:
TaskList class: Add print contents functionality

derekjxtan committed:
TaskList class: Add print tasks on a date functionality

derekjxtan committed:
TaskList class: Add mark done functionality

derekjxtan committed:
TaskList class: Add unmark as done functionality

derekjxtan committed:
mend

TaskList class: Add print statements to mark function

derekjxtan committed:
mend

TaskList class: Add print statements to unmark function

derekjxtan committed:
mend

TaskList class: Add functionality to check if index is valid

derekjxtan committed:
mend

TaskList class: Add index valid checks to functions

derekjxtan committed:
Duke class: Update to use TaskList class

derekjxtan committed:
Add Storage class

derekjxtan committed:
Storage class: Add implementation for saveData function

derekjxtan committed:
Storage class: Add implementation for loadData function

derekjxtan committed:
TaskList class: Update to use Storage class

derekjxtan committed:
Add UI class

derekjxtan committed:
UI class: implement print welcome message functionality

derekjxtan committed:
UI class: implement print Goodbye message functionality

derekjxtan committed:
UI class: implement print task added, deleted, marked, unmarked functionality

derekjxtan committed:
Duke class: update to use UI class

derekjxtan committed:
TaskList class: Move output statements to UI class

derekjxtan committed:
TaskList class: Move print tasks on functionality to UI class

derekjxtan committed:
Refactor Duke class

derekjxtan committed:
Add packages

derekjxtan committed:
Merge branch 'branch-more-oop'

derekjxtan committed:
Merge branch 'branch-packages'

derekjxtan committed:
Merge branch 'add-gradle-support'

derekjxtan committed:
Fix build.gradle configurations

derekjxtan committed:
Add ParserTest class

derekjxtan committed:
Add TaskListTest class

derekjxtan committed:
Merge branch 'branch-junit'

derekjxtan committed:
Storage class: Add JavaDocs

derekjxtan committed:
Parser class: Add JavaDocs

derekjxtan committed:
TaskList class: Add JavaDocs

derekjxtan committed:
UI class: Add JavaDocs

derekjxtan committed:
Task class: Add JavaDocs

derekjxtan committed:
ToDo class: Add JavaDocs

derekjxtan committed:
Deadline class: Add JavaDocs

derekjxtan committed:
Event class: Add JavaDocs

derekjxtan committed:
Task class: Rename done to isDone

derekjxtan committed:
Task class: Add containsKeyword function

derekjxtan committed:
TaskList class: Add getTasksContainingKeyword function

derekjxtan committed:
Duke class: Add functionality to search by keyword

derekjxtan committed:
Merge branch 'branch-level-9'

derekjxtan committed:
Merge branch 'master' into branch-A-javaDoc

derekjxtan committed:
Add JavaDocs for newly added commits

derekjxtan committed:
Merge branch 'branch-A-javaDoc'

derekjxtan committed:
Merge branch 'master' into branch-A-CodingStandard

derekjxtan committed:
Merge branch 'branch-A-CodingStandard'

derekjxtan committed:
bug fix: Create directory and file to save data

derekjxtan committed:
bug fix: Save tasks to file on change

derekjxtan committed:
Move read user input functionality to UI class

derekjxtan committed:
UI class: Remove datatime formatter

derekjxtan committed:
Remove Command enum

derekjxtan committed:
Add Command class

derekjxtan committed:
Add AddCommand class

derekjxtan committed:
Add DeleteCommand class

derekjxtan committed:
Add ExitCommand class

derekjxtan committed:
Add MarkCommand class

derekjxtan committed:
Add UnmarkCommand class

derekjxtan committed:
Add ListCommand class

derekjxtan committed:
Add OnCommand class

derekjxtan committed:
Add FindCommand class

derekjxtan committed:
Rename FinaCommand file to FindCommand

derekjxtan committed:
Duke class: Update to use Command class

derekjxtan committed:
Command class: Add isExit function

derekjxtan committed:
Move Command classes to Command package

derekjxtan committed:
Rename Command package to command

derekjxtan committed:
Add javadocs to command classes

derekjxtan committed:
Add exception throwing to constructors in command classes

derekjxtan committed:
Command class: Override equals method

derekjxtan committed:
ParserTest class: Update tests

derekjxtan committed:
Merge branch 'command-abstraction'

derekjxtan committed:
Add checkstyle config files

derekjxtan committed:
Update build.gradle for checkstyle

derekjxtan committed:
ParserTest.java: Fix checkstyle violations

derekjxtan committed:
TaskListTest.java: Fix checkstyle violations

derekjxtan committed:
Fix checkstyle violations

derekjxtan committed:
Merge branch 'branch-A-CheckStyle'

derekjxtan committed:
Update build.gradle

derekjxtan committed:
Duke class: Update to extend from javafx Application

derekjxtan committed:
Duke class: Add basic GUI

derekjxtan committed:
Update GUI to follow mockup layout

ncmathan commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,59 @@
+     * @return A string representing the event to be saved.
+     */
+    @Override
+    public String toSaveStateString() {
```
A better method name might be `formatStringToSave`

ncmathan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,113 @@
+    public static Command parseUserInput(String input) throws InvalidCommandException {
+        ArrayList<String> parsedInput= new ArrayList<>();
+
+        String[] splitInputByDateDelimiter = input.split(DELIMITER_DATE);
```
A better variable name might be `splitInputs`

ncmathan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,113 @@
+
+        String commandString = splitInputByDateDelimiter[0];
+        String[] splitCommandStringByCommandDelimiter = commandString.split(DELIMITER_COMMAND);
+        String command = splitCommandStringByCommandDelimiter[0];
```
a better variable name might be `splitCommands`

ncmathan reviewed (commented):
Overall great naming convention followed. Coding standards are also followed well. Keep it up! 👍

derekjxtan committed:
Update application to use fxml for GUI

derekjxtan committed:
Duke class: Remove unused functions

derekjxtan committed:
Duke class: Fix implementation of getResponse

derekjxtan committed:
Fix CheckStyle violations

juliusgambe commented on src/main/java/command/Command.java:
```diff
@@ -0,0 +1,49 @@
+package command;
```
If I'm not wrong, the sub-packages should still be under `Duke` :D

juliusgambe commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,49 @@
+
+    }
+
+    public void run() {
+        ui.printWelcomeMessage();
+        String input;
+
+        while (true) {
+            input = ui.readCommand();
+            try {
+                Command command = Parser.parseUserInput(input);
+                command.execute(tasks, ui, storage);
+                if (command.isExit()) {
+                    break;
+                }
+            } catch (DukeException e) {
+                System.out.println(e.getMessage());
+            }
+        }
+    }
```
You should add javadocs for public classes and methods : D

juliusgambe commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,190 @@
+    /**
+     * Prints all tasks in the list.
+     */
+    public void printContents() {
```
Perhaps a better name is `printTasks`

juliusgambe commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,190 @@
+     *
+     * @param storage storage object to save the tasks.
+     */
+    public void saveState(Storage storage) throws StorageException {
+        ArrayList<String> stringRepresentation = new ArrayList<>();
+        for (int i = 0; i < this.tasks.size(); i++) {
+            stringRepresentation.add(this.tasks.get(i).toSaveStateString());
+        }
+        storage.saveData(stringRepresentation);
+    }
+
+    /**
+     * Loads the tasks from a file.
+     *
+     * @param storedData data stored in a file.
+     */
+    private void loadState(ArrayList<String> storedData) {
+        for (int i = 0; i < storedData.size(); i++) {
+            String[] taskArray = storedData.get(i).split(" / ");
+            Task task;
+
+            if (taskArray[0].equals("todo")) {
+                task = new ToDo(taskArray[2]);
+            } else if (taskArray[0].equals("deadline")) {
+                task = new Deadline(taskArray[2], LocalDateTime.parse(taskArray[3], dateTimeInputFormatter));
+            } else {
+                task = new Event(taskArray[2], LocalDateTime.parse(taskArray[3], dateTimeInputFormatter),
+                        LocalDateTime.parse(taskArray[4], dateTimeInputFormatter));
+            }
+
+            if (taskArray[1].equals("1")) {
+                task.markAsDone();
+            }
+            this.tasks.add(task);
+        }
+        System.out.println("Successfully loaded saved state");
+    }
```
You may want to consider abstracting these methods to your `Storage` class

juliusgambe commented on src/main/java/duke/ToDo.java:
```diff
@@ -0,0 +1,45 @@
+
+import java.time.LocalDate;
+
+public class ToDo extends Task{
```
You may want to follow the naming convention of `ToDoTask` etc. like you did for `Command`

juliusgambe reviewed (commented):
LGTM. I think there's some small changes to be made in the naming and adding of javadocs. :D

derekjxtan committed:
Add welcome message to GUI

derekjxtan committed:
Bug fix: Fix long responses automatically truncated

derekjxtan committed:
Remove CLI UI features and associated functionality

derekjxtan committed:
Bug Fix: Close Application when 'bye' command is issued

derekjxtan committed:
MainWindow class: Add javadoc comment for addWelcomeMessage function

derekjxtan committed:
Merge branch 'branch-Level-10'

derekjxtan committed:
Update mainClass in build.gradle

derekjxtan committed:
Add assertions

Assumptions are made in the code logic, for example from and to of Event
is assumed to be not null in the constructor.

These assumptions should be eliminated using assertions to prevent
related bugs.

Assertions have been added to:
* Task
* Deadline
* Event
* Storage
* TaskList
* Duke

derekjxtan committed:
Merge pull request #1 from derekjxtan/branch-A-Assertions

Add assertions

derekjxtan committed:
Remove used methods

derekjxtan committed:
Add @Test tag to JUnit tests

derekjxtan committed:
Task class: rename toSaveStateString to formatStringToSave

derekjxtan committed:
Add Savable interface

The storage class takes in an ArrayList of Strings to save to file.

Does not fully abstract out the functionality to saving objects to
file as it requires the object to be converted into a string before
being passed to saveData in the Storage class.

Adding a Savable interface to be implemented by all classes that can be
saved to file allows for full abstraction.

derekjxtan committed:
Fix checkstyle violations

derekjxtan committed:
Merge pull request #2 from derekjxtan/branch-A-CodeQuality

Branch a code quality

derekjxtan committed:
Setup continuous integration

derekjxtan committed:
Add StatisticsCommand class

derekjxtan committed:
Merge pull request #3 from derekjxtan/A-CI

Setup continuous integration

derekjxtan committed:
Parser class: Add creation of StatisticsCommand

derekjxtan committed:
TaskList class: Add functionality to get number of tasks completed in the past n days

derekjxtan committed:
Implement getStatistics functionality

derekjxtan committed:
Fix checkstyle violations

derekjxtan committed:
Merge pull request #4 from derekjxtan/branch-C-Statistics

Branch c statistics

derekjxtan committed:
Update README

derekjxtan committed:
Merge pull request #6 from derekjxtan/branch-A-UserGuide

Add user guide

derekjxtan committed:
Tasklist class: update comments

derekjxtan committed:
Update README.md

derekjxtan committed:
Update README.md

derekjxtan committed:
Update README.md

derekjxtan committed:
Update build.gradle mac dependencies for javafx
